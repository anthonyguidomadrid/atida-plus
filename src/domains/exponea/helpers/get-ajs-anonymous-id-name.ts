export const getAjsAnonymousIdName = (): string => 'ajs_anonymous_id'
