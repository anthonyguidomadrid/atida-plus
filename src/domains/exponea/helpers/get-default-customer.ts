import getConfig from 'next/config'

export const getDefaultCustomer = (): string => {
  const { serverRuntimeConfig } = getConfig()
  return serverRuntimeConfig.exponea.defaultCustomer
}
