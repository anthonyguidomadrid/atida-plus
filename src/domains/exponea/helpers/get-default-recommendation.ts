import getConfig from 'next/config'
export const getDefaultRecommendation = (
  locale: string | undefined
): string => {
  const { serverRuntimeConfig } = getConfig()
  return serverRuntimeConfig.exponea.defaultRecommendation[locale as string]
}
