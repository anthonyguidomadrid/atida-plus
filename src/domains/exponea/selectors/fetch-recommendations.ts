import { createSelector } from '@reduxjs/toolkit'
import { RootState } from '~domains/redux'

const selectRecommendations = (state: RootState) => state?.client?.exponea

export const selectIsLoading = createSelector(
  selectRecommendations,
  recommendedProducts =>
    recommendedProducts?.recommendations?.isLoading ?? false
)

export const selectRecommendedProducts = createSelector(
  selectRecommendations,
  recommendedProducts => recommendedProducts?.recommendations?.details
)
