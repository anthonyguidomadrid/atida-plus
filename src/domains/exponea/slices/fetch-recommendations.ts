import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { clearStateErrors } from '~domains/redux/actions'
import {
  ExponeaFetchRecommendationsDataState,
  ExponeaRecommendationPayload,
  FetchRecommendationResponse
} from '~domains/exponea/types'

const initialState: ExponeaFetchRecommendationsDataState = {
  isLoading: false,
  wasSuccess: false,
  wasError: false
}

export const fetchRecommendationsSlice = createSlice({
  name: 'recommendations',
  initialState,
  reducers: {
    trigger(_state, _action: PayloadAction<ExponeaRecommendationPayload>) {
      // no-op, used for triggering saga
    },
    request(state) {
      state.isLoading = true
      state.wasSuccess = false
      state.wasError = false
      delete state.error
    },
    success(state, action: PayloadAction<FetchRecommendationResponse>) {
      let updated = false
      const data = action.payload
      if (typeof data !== 'undefined') {
        if (typeof state.details !== 'undefined') {
          state.details.map(value => {
            if (value.recommendationId == data.recommendationId) {
              value.items = data.items
              value.isPersonalized = data.isPersonalized
              updated = true
            }
          })
          !updated && state.details.push(data)
        } else {
          state.details = [data]
        }
      }
      state.wasSuccess = true
    },
    failure(state, action: PayloadAction<{ message: string }>) {
      const { message } = action.payload
      state.wasError = true
      state.error = message
    },
    fulfill(state) {
      state.isLoading = false
    }
  },
  extraReducers: builder => {
    builder.addCase(clearStateErrors, (state, _action) => {
      state.wasError = false
      delete state.error
    })
  }
})
