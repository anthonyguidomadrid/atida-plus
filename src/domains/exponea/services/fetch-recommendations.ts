import { createClient, hasOwnProperty, logger } from '~helpers'
import getConfig from 'next/config'
import { sha256 } from 'js-sha256'
import {
  ChannelPrice,
  ElasticSearchProduct,
  normalizeElasticSearchProduct
} from '~domains/product'
import {
  createElasticSearchClient,
  getElasticSearchProductIndex
} from '~domains/elasticsearch/helpers/client'
import {
  ExponeaRecommendationResponse,
  ExponeaResult,
  FetchRecommendationResponse,
  SuccessfulExponeaResult
} from '~domains/exponea/types'
import { getDefaultCustomer } from '~domains/exponea/helpers/get-default-customer'
import { getDefaultRecommendation } from '~domains/exponea/helpers/get-default-recommendation'
import { DEFAULT_PRODUCT_RECOMMENDATIONS_QUANTITY } from '~config/constants/recommendations'
import { SessionChannelType } from '~domains/basket/types'
import { FeatureFlag } from '~config/constants/feature-flags'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

const responseHasRecommendations = (
  result: ExponeaResult
): result is SuccessfulExponeaResult => hasOwnProperty(result, 'value')

export const fetchRecommendations = async (
  locale: string,
  cookie?: string,
  data?: {
    email?: string
    recommendationId?: string
    productId?: string
    categoryId?: string
    itemsQuantity?: number
    sessionChannel?: SessionChannelType
  }
): Promise<FetchRecommendationResponse> => {
  const { serverRuntimeConfig } = getConfig()

  const projectToken = serverRuntimeConfig.exponea.projectToken[locale]
  const apiKey = serverRuntimeConfig.exponea.apiKey[locale]
  const apiSecret = serverRuntimeConfig.exponea.apiSecret[locale]

  if (!projectToken) {
    logger.warn(`Exponea project token not available for locale: ${locale}`)
  }

  if (!apiKey) {
    logger.warn(`Exponea API key not available for locale: ${locale}`)
  }

  if (!apiSecret) {
    logger.warn(`Exponea API secret not available for locale: ${locale}`)
  }

  const auth = Buffer.from(`${apiKey}:${apiSecret}`, 'binary').toString(
    'base64'
  )

  const client = createClient({
    options: {
      headers: {
        Authorization: `Basic ${auth}`
      }
    },
    locale
  })

  const logPayload = {
    recommendationId: data?.recommendationId,
    categoryId: data?.categoryId,
    productId: data?.productId
  }

  const itemsQuantity =
    data?.itemsQuantity || DEFAULT_PRODUCT_RECOMMENDATIONS_QUANTITY

  try {
    const exponeaResponse = await client.post<ExponeaRecommendationResponse>(
      `https://api.exponea.com/data/v2/projects/${projectToken}/customers/attributes`,
      {
        customer_ids: {
          email_id: data?.email
            ? sha256(data?.email.trim().toLowerCase())
            : cookie
            ? undefined
            : getDefaultCustomer(), // Identification for default recommendations when no cookie/email available
          cookie: cookie?.replace(/['"]+/g, '') ?? undefined
        },
        attributes: [
          {
            type: 'recommendation',
            // Default recommendation if the passed doesn't exist
            id: data?.recommendationId ?? getDefaultRecommendation(locale),
            fillWithRandom: true,
            size: itemsQuantity,
            catalogAttributesWhitelist: ['item_id', 'product_id'],
            items: data?.productId && { [`${data?.productId}`]: 1 },
            catalogFilter: data?.categoryId && [
              {
                property: 'web_category',
                constraint: {
                  type: 'string',
                  operator: 'contains',
                  operands: [{ type: 'constant', value: data?.categoryId }]
                }
              }
            ]
          }
        ]
      },
      { timeout: serverRuntimeConfig.exponea.timeout, baseURL: '' }
    )

    const result = exponeaResponse.data.results[0] as ExponeaResult
    const exponeaProducts = responseHasRecommendations(result)
      ? result.value
      : []

    const recommendedProducts = exponeaProducts.map(({ item_id }) => item_id)

    const isPersonalized =
      exponeaProducts.filter(
        ({ recommendation_source, engine_name }) =>
          recommendation_source === 'model_personalized' ||
          engine_name === 'Custom' // Used when we create manual selection recommendation in Exponea
      ).length > 0

    if (recommendedProducts.length === 0) {
      const error = !responseHasRecommendations(result) ? result.error : null

      if (error) {
        logger.error(
          { error, ...logPayload },
          `Exponea did not return any recommendations for ID: ${data?.recommendationId}`
        )
      } else {
        logger.info(
          logPayload,
          `Exponea did not return any recommendations for ID: ${data?.recommendationId} but did not return an error`
        )
      }

      return {
        recommendationId: data?.recommendationId,
        items: []
      }
    }

    const elasticSearchClient = createElasticSearchClient()
    const elasticSearchResponse = await elasticSearchClient.mget<{
      docs: ElasticSearchProduct[]
    }>({
      index: getElasticSearchProductIndex(locale),
      body: {
        ids: recommendedProducts
      }
    })

    const onlyFound = elasticSearchResponse.body.docs.filter(doc => doc.found)
    const items =
      onlyFound.length < 4
        ? []
        : onlyFound.length >= 4 && onlyFound.length < 6
        ? onlyFound.slice(0, 4)
        : onlyFound.length >= 6 && onlyFound.length < 8
        ? onlyFound.slice(0, 6)
        : onlyFound.length >= 8 && onlyFound.length < 12
        ? onlyFound.slice(0, 8)
        : onlyFound.slice(0, 12)

    // TODO: Remove when the prices channel on products are ready to be used
    const mockedProductPriceChannels = (await loadFeatureFlag(
      locale,
      FeatureFlag.MUTLICHANNEL_PRICES_MOCKED_PRODUCT_PRICE_CHANNEL
    )) as { channel_prices: ChannelPrice[] }
    return {
      recommendationId: data?.recommendationId,
      isPersonalized,
      items: items.map(doc =>
        normalizeElasticSearchProduct(
          locale,
          doc,
          data?.sessionChannel,
          mockedProductPriceChannels?.channel_prices
        )
      )
    }
  } catch (error) {
    return {
      recommendationId: data?.recommendationId,
      items: []
    }
  }
}
