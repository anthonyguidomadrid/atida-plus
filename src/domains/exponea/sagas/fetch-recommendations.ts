import type { PayloadAction } from '@reduxjs/toolkit'
import type { SagaIterator } from 'redux-saga'
import {
  all,
  call,
  getContext,
  put,
  select,
  takeEvery
} from 'redux-saga/effects'
import { createClient, logger } from '~helpers'
import {
  fetchRecommendationsTrigger,
  fetchRecommendationsRequest,
  fetchRecommendationsSuccess,
  fetchRecommendationsFailure,
  fetchRecommendationsFulfill
} from '../slices'
import { selectCustomerDetails } from '~domains/account/selectors/customer'
import { ExponeaRecommendationPayload } from '~domains/exponea/types'
import { getErrorMessage } from '~helpers/error'
import { StrictEffect } from '@redux-saga/types'

function* doFetchRecommendationsSaga({
  payload
}: PayloadAction<ExponeaRecommendationPayload>): Generator<
  StrictEffect,
  void,
  any
> {
  try {
    const customerDetails = yield select(selectCustomerDetails)
    yield put(fetchRecommendationsRequest())
    const locale = yield getContext('locale')
    const client = createClient({ locale })
    const response = yield call(client.post, '/api/recommendations/fetch', {
      email: customerDetails ? customerDetails.email : undefined,
      recommendationId: payload.recommendationId,
      productId: payload.productId ?? undefined,
      categoryId: payload.categoryId ?? undefined,
      itemsQuantity: payload.itemsQuantity ?? undefined
    })
    yield put(fetchRecommendationsSuccess(response.data))
    // This might be needed in the future for analytics TODO: Send recommendation event to segment
    // yield put(triggerReportRecommendation())
  } catch (error) {
    const errorMessages = getErrorMessage(error)
    logger.error(errorMessages)
    if (Array.isArray(errorMessages)) {
      yield put(
        fetchRecommendationsFailure({
          message: errorMessages[0]
        })
      )
    } else {
      yield put(
        fetchRecommendationsFailure({
          message: errorMessages
        })
      )
    }
  } finally {
    yield put(fetchRecommendationsFulfill())
  }
}

export function* fetchRecommendationsSaga(): SagaIterator {
  yield all([
    takeEvery(fetchRecommendationsTrigger, doFetchRecommendationsSaga)
  ])
}
