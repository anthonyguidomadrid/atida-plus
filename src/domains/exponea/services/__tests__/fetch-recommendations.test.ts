/**
 * @jest-environment node
 */
// @ts-ignore
import { mget } from '@elastic/elasticsearch'
import { elasticsearchProducts } from '~domains/basket/__mocks__/basket'
import {
  ExponeaRecommendationResponse,
  recommendedProducts
} from '~domains/exponea/__mocks__/recommendation'
import { fetchRecommendations } from '../fetch-recommendations'
import axios from 'axios'
import { getDefaultCustomer } from '~domains/exponea/helpers/get-default-customer'
import { getDefaultRecommendation } from '~domains/exponea/helpers/get-default-recommendation'

describe(fetchRecommendations, () => {
  beforeEach(() => {
    mget.mockResolvedValue(elasticsearchProducts)
  })
  beforeEach(() => {
    ;(axios.post as jest.Mock).mockResolvedValue(ExponeaRecommendationResponse)
  })

  const data = {
    email: 'someCustomerEmail',
    recommendationId: 'someRecommendationId',
    productId: 'someProductId',
    categoryId: 'someCategoryId',
    sessionChasessionChannel: { sku: 'some-sku', channel: 'amazon' }
  }

  it('passes the correct parameters to exponea and elasticsearch request', async () => {
    await fetchRecommendations('en-gb', 'someExponeaCookie', data)

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-GB/customers/attributes',
      {
        customer_ids: {
          email_id:
            '834e02be69d5b35337630c0bfad5236643604208e94539f72a346b7fbc28f527', // Base64 hashed email
          cookie: 'someExponeaCookie'
        },
        attributes: [
          {
            type: 'recommendation',
            id: data?.recommendationId,
            fillWithRandom: true,
            size: 8,
            catalogAttributesWhitelist: ['item_id', 'product_id'],
            items: { [`${data?.productId}`]: 1 },
            catalogFilter: [
              {
                constraint: {
                  operands: [
                    {
                      type: 'constant',
                      value: 'someCategoryId'
                    }
                  ],
                  operator: 'contains',
                  type: 'string'
                },
                property: 'web_category'
              }
            ]
          }
        ]
      },
      expect.any(Object)
    )
    expect(mget).toHaveBeenCalledWith({
      body: {
        ids: ['100000001', '100000002', '100000003', '100000004']
      },
      index: 'index-en-gb'
    })
  })

  it('passes the correct parameters to exponea and elasticsearch request when email is missing', async () => {
    await fetchRecommendations('en-gb', 'someExponeaCookie', {
      recommendationId: data.recommendationId,
      productId: data.productId,
      categoryId: data.categoryId
    })

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-GB/customers/attributes',
      {
        customer_ids: {
          cookie: 'someExponeaCookie'
        },
        attributes: [
          {
            type: 'recommendation',
            id: data?.recommendationId,
            fillWithRandom: true,
            size: 8,
            catalogAttributesWhitelist: ['item_id', 'product_id'],
            items: { [`${data?.productId}`]: 1 },
            catalogFilter: [
              {
                constraint: {
                  operands: [
                    {
                      type: 'constant',
                      value: 'someCategoryId'
                    }
                  ],
                  operator: 'contains',
                  type: 'string'
                },
                property: 'web_category'
              }
            ]
          }
        ]
      },
      expect.any(Object)
    )
    expect(mget).toHaveBeenCalledWith({
      body: {
        ids: ['100000001', '100000002', '100000003', '100000004']
      },
      index: 'index-en-gb'
    })
  })

  it('passes the correct parameters to exponea and elasticsearch request when cookie is missing', async () => {
    await fetchRecommendations('en-gb', undefined, data)

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-GB/customers/attributes',
      {
        customer_ids: {
          email_id:
            '834e02be69d5b35337630c0bfad5236643604208e94539f72a346b7fbc28f527' // Hashed email
        },
        attributes: [
          {
            type: 'recommendation',
            id: data?.recommendationId,
            fillWithRandom: true,
            size: 8,
            catalogAttributesWhitelist: ['item_id', 'product_id'],
            items: { [`${data?.productId}`]: 1 },
            catalogFilter: [
              {
                constraint: {
                  operands: [
                    {
                      type: 'constant',
                      value: 'someCategoryId'
                    }
                  ],
                  operator: 'contains',
                  type: 'string'
                },
                property: 'web_category'
              }
            ]
          }
        ]
      },
      expect.any(Object)
    )
    expect(mget).toHaveBeenCalledWith({
      body: {
        ids: ['100000001', '100000002', '100000003', '100000004']
      },
      index: 'index-en-gb'
    })
  })

  it('passes the correct parameters to exponea and elasticsearch request when email and cookie is missing', async () => {
    await fetchRecommendations('en-gb', 'someExponeaCookie', {
      recommendationId: data.recommendationId,
      email: data.email
    })

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-GB/customers/attributes',
      {
        customer_ids: {
          cookie: 'someExponeaCookie',
          email_id:
            '834e02be69d5b35337630c0bfad5236643604208e94539f72a346b7fbc28f527'
        },
        attributes: [
          {
            type: 'recommendation',
            id: data?.recommendationId,
            fillWithRandom: true,
            size: 8,
            catalogAttributesWhitelist: ['item_id', 'product_id']
          }
        ]
      },
      expect.any(Object)
    )
  })

  it('passes the correct parameters to exponea and elasticsearch request when cookie and email is missing', async () => {
    await fetchRecommendations('en-gb', undefined, {
      recommendationId: data.recommendationId,
      productId: data.productId,
      categoryId: data.categoryId
    })

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-GB/customers/attributes',
      {
        customer_ids: {
          email_id: getDefaultCustomer()
        },
        attributes: [
          {
            type: 'recommendation',
            id: data?.recommendationId,
            fillWithRandom: true,
            size: 8,
            catalogAttributesWhitelist: ['item_id', 'product_id'],
            items: { [`${data?.productId}`]: 1 },
            catalogFilter: [
              {
                constraint: {
                  operands: [
                    {
                      type: 'constant',
                      value: 'someCategoryId'
                    }
                  ],
                  operator: 'contains',
                  type: 'string'
                },
                property: 'web_category'
              }
            ]
          }
        ]
      },
      expect.any(Object)
    )
    expect(mget).toHaveBeenCalledWith({
      body: {
        ids: ['100000001', '100000002', '100000003', '100000004']
      },
      index: 'index-en-gb'
    })
  })

  it('passes the correct parameters to exponea and elasticsearch request when recommendationId is missing', async () => {
    await fetchRecommendations('en-gb', 'someExponeaCookie', {
      email: data.email,
      productId: data.productId,
      categoryId: data.categoryId
    })

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-GB/customers/attributes',
      {
        customer_ids: {
          email_id:
            '834e02be69d5b35337630c0bfad5236643604208e94539f72a346b7fbc28f527',
          cookie: 'someExponeaCookie'
        },
        attributes: [
          {
            type: 'recommendation',
            id: getDefaultRecommendation('en-gb'),
            fillWithRandom: true,
            size: 8,
            catalogAttributesWhitelist: ['item_id', 'product_id'],
            items: { [`${data?.productId}`]: 1 },
            catalogFilter: [
              {
                constraint: {
                  operands: [
                    {
                      type: 'constant',
                      value: 'someCategoryId'
                    }
                  ],
                  operator: 'contains',
                  type: 'string'
                },
                property: 'web_category'
              }
            ]
          }
        ]
      },
      expect.any(Object)
    )
    expect(mget).toHaveBeenCalledWith({
      body: {
        ids: ['100000001', '100000002', '100000003', '100000004']
      },
      index: 'index-en-gb'
    })
  })

  it('uses the correct exponea project token for a different locale', async () => {
    await fetchRecommendations('pt-pt', 'someExponeaCookie', data)

    expect(axios.post).toHaveBeenCalledWith(
      'https://api.exponea.com/data/v2/projects/someProjectToken-PT/customers/attributes',
      expect.any(Object),
      expect.any(Object)
    )
  })

  it('returns the normalized recommendation response with products', async () => {
    const response = await fetchRecommendations(
      'en-gb',
      'someExponeaCookie',
      data
    )
    expect(response).toStrictEqual(recommendedProducts)
  })

  it('returns an empty list when exponea does not return any recommendations', async () => {
    ;(axios.post as jest.Mock).mockResolvedValueOnce({
      data: {
        results: [
          {
            error: 'No permission',
            success: false
          }
        ],
        success: true
      }
    })
    const response = await fetchRecommendations(
      'en-gb',
      'someExponeaCookie',
      data
    )
    expect(response).toMatchSnapshot()
  })

  it('returns an empty list when exponea API fails', async () => {
    ;(axios.post as jest.Mock).mockRejectedValueOnce({})
    const response = await fetchRecommendations(
      'en-gb',
      'someExponeaCookie',
      data
    )
    expect(response).toMatchSnapshot()
  })
})
