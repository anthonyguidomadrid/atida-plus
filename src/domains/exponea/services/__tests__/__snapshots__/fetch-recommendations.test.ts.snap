// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchRecommendations returns an empty list when exponea API fails 1`] = `
Object {
  "items": Array [],
  "recommendationId": "someRecommendationId",
}
`;

exports[`fetchRecommendations returns an empty list when exponea does not return any recommendations 1`] = `
Object {
  "items": Array [],
  "recommendationId": "someRecommendationId",
}
`;
